File: src/shared/native-backend/fs.ts
import type { NativeProjectId, NativeRelativePath, NativeWorktreeId } from "./ids";

export type NativeFsEntryKind = "directory" | "file" | "other" | "symlink";
export type NativeFsEntryStatus =
    | "clean"
    | "conflicted"
    | "created"
    | "deleted"
    | "ignored"
    | "modified"
    | "renamed"
    | "unknown";
export type NativeFsMutationOrigin = "agent" | "external" | "system" | "user";
export type NativeFsVisibilityPolicy =
    | "hidden_by_policy"
    | "noisy"
    | "permission_denied"
    | "special"
    | "too_large_to_expand"
    | "visible";

export type NativeFsEntry = {
    readonly path: string;
    readonly relativePath: NativeRelativePath;
    readonly projectId: NativeProjectId;
    readonly worktreeId: NativeWorktreeId | null;
    readonly kind: NativeFsEntryKind;
    readonly isDirectory: boolean;
    readonly isSymlink: boolean;
    readonly isBinary: boolean;
    readonly isTooLarge: boolean;
    readonly sizeBytes: number | null;
    readonly mtimeMs: number | null;
    readonly contentHash: string | null;
    readonly status: NativeFsEntryStatus;
    readonly visibility?: NativeFsVisibilityPolicy | null;
};

export type NativeFsReadFileInput = {
    readonly projectId: NativeProjectId;
    readonly worktreeId: NativeWorktreeId | null;
    readonly relativePath: NativeRelativePath;
    readonly maxBytes?: number | null;
};

export type NativeFsReadFileResult = {
    readonly projectId: NativeProjectId;
    readonly worktreeId: NativeWorktreeId | null;
    readonly path: string;
    readonly relativePath: NativeRelativePath;
    readonly name?: string | null;
    readonly content: string | null;
    readonly encoding: string | null;
    readonly lineEnding: string | null;
    readonly contentHash: string | null;
    readonly sizeBytes: number;
    readonly mtimeMs: number;
    readonly mimeType?: string | null;
    readonly kind?: "binary" | "image" | "text" | string | null;
    readonly imageDataBase64?: string | null;
    readonly isBinary: boolean;
    readonly isTooLarge: boolean;
};

export type NativeFsWriteFileInput = {
    readonly projectId: NativeProjectId;
    readonly worktreeId: NativeWorktreeId | null;
    readonly relativePath: NativeRelativePath;
    readonly content: string;
    readonly expectedContentHash: string | null;
    readonly expectedModifiedAtMs?: number | null;
    readonly origin: NativeFsMutationOrigin;
};

export type NativeFsConflict = {
    readonly reason: string;
    readonly currentContentHash: string | null;
    readonly externalMtimeMs: number | null;
};

export type NativeFsWriteFileResult = {
    readonly entry: NativeFsEntry;
    readonly conflict: NativeFsConflict | null;
    readonly file?: NativeFsReadFileResult | null;
};

export type NativeFsCreateEntryInput = {
    readonly projectId: NativeProjectId;
    readonly worktreeId: NativeWorktreeId | null;
    readonly parentRelativePath: NativeRelativePath | null;
    readonly name: string;
    readonly kind: NativeFsEntryKind;
    readonly origin: NativeFsMutationOrigin;
};

export type NativeFsRenameEntryInput = {
    readonly projectId: NativeProjectId;
    readonly worktreeId: NativeWorktreeId | null;
    readonly relativePath: NativeRelativePath;
    readonly nextName: string;
    readonly nextParentRelativePath?: NativeRelativePath | null;
    readonly origin: NativeFsMutationOrigin;
};

export type NativeFsDeleteEntryInput = {
    readonly projectId: NativeProjectId;
    readonly worktreeId: NativeWorktreeId | null;
    readonly relativePath: NativeRelativePath;
    readonly origin: NativeFsMutationOrigin;
};

export type NativeFsCopyEntriesInput = {
    readonly projectId: NativeProjectId;
    readonly worktreeId: NativeWorktreeId | null;
    readonly sourceRelativePaths: readonly NativeRelativePath[];
    readonly destinationParentRelativePath: NativeRelativePath | null;
    readonly origin: NativeFsMutationOrigin;
};

export type NativeFsCopyExternalEntriesInput = {
    readonly projectId: NativeProjectId;
    readonly worktreeId: NativeWorktreeId | null;
    readonly sourcePaths: readonly string[];
    readonly destinationParentRelativePath: NativeRelativePath | null;
    readonly origin: NativeFsMutationOrigin;
};

export type NativeFsEntryMutationResult = {
    readonly kind: NativeFsEntryKind;
    readonly name: string;
    readonly parentRelativePath: NativeRelativePath | null;
    readonly relativePath: NativeRelativePath;
    readonly entry?: NativeFsEntry | null;
};

export type NativeFsEntryMutationListResult = {
    readonly entries: readonly NativeFsEntryMutationResult[];
};

export type NativeFsRecordExternalMutationInput = {
    readonly projectId: NativeProjectId;
    readonly worktreeId: NativeWorktreeId | null;
    readonly relativePaths: readonly NativeRelativePath[];
    readonly origin: NativeFsMutationOrigin;
};

export type NativeFsRevealEntryInfoInput = {
    readonly projectId: NativeProjectId;
    readonly worktreeId: NativeWorktreeId | null;
    readonly relativePath: NativeRelativePath | null;
};

export type NativeFsRevealEntryInfoResult = {
    readonly projectId: NativeProjectId;
    readonly worktreeId: NativeWorktreeId | null;
    readonly path: string;
    readonly relativePath: NativeRelativePath | null;
    readonly exists: boolean;
    readonly kind: NativeFsEntryKind;
};

export type NativeFsWatchInput = {
    readonly projectId: NativeProjectId;
    readonly worktreeId: NativeWorktreeId | null;
};

export type NativeProjectTreeInvalidation = {
    readonly projectId: NativeProjectId;
    readonly worktreeId: NativeWorktreeId | null;
    readonly relativePaths: readonly NativeRelativePath[] | null;
    readonly occurredAt: string;
};

export type NativeFsWatchEvent = {
    readonly projectId: NativeProjectId;
    readonly worktreeId: NativeWorktreeId | null;
    readonly relativePath: NativeRelativePath | null;
    readonly kind: string;
    readonly origin: NativeFsMutationOrigin;
    readonly occurredAt: string;
};

export type NativeOpenBufferState = {
    readonly path: string;
    readonly relativePath: NativeRelativePath;
    readonly projectId: NativeProjectId;
    readonly worktreeId: NativeWorktreeId | null;
    readonly contentHash: string | null;
    readonly isDirty: boolean;
};

export type NativePathPolicy = {
    readonly path: string;
    readonly visible: boolean;
    readonly aiAccess: boolean;
    readonly reason: string | null;
};

File: crates/comando-types/src/fs.rs
use serde::{Deserialize, Serialize};

use crate::ids::{ProjectId, RelativePath, WorktreeId};

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum NativeFsEntryKind {
    File,
    Directory,
    Symlink,
    Other,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum NativeFsEntryStatus {
    Clean,
    Modified,
    Created,
    Deleted,
    Renamed,
    Conflicted,
    Ignored,
    Unknown,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum NativeFsMutationOrigin {
    User,
    Agent,
    External,
    System,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum NativeFsVisibilityPolicy {
    Visible,
    Noisy,
    Special,
    HiddenByPolicy,
    TooLargeToExpand,
    PermissionDenied,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct NativeFsEntry {
    pub path: String,
    pub relative_path: RelativePath,
    pub project_id: ProjectId,
    pub worktree_id: Option<WorktreeId>,
    pub kind: NativeFsEntryKind,
    pub is_directory: bool,
    pub is_symlink: bool,
    pub is_binary: bool,
    pub is_too_large: bool,
    pub size_bytes: Option<u64>,
    pub mtime_ms: Option<u64>,
    pub content_hash: Option<String>,
    pub status: NativeFsEntryStatus,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub visibility: Option<NativeFsVisibilityPolicy>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct NativeFsReadFileInput {
    pub project_id: ProjectId,
    pub worktree_id: Option<WorktreeId>,
    pub relative_path: RelativePath,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub max_bytes: Option<u64>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct NativeFsReadFileResult {
    pub project_id: ProjectId,
    pub worktree_id: Option<WorktreeId>,
    pub path: String,
    pub relative_path: RelativePath,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    pub content: Option<String>,
    pub encoding: Option<String>,
    pub line_ending: Option<String>,
    pub content_hash: Option<String>,
    pub size_bytes: u64,
    pub mtime_ms: u64,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub mime_type: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub kind: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub image_data_base64: Option<String>,
    pub is_binary: bool,
    pub is_too_large: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct NativeFsWriteFileInput {
    pub project_id: ProjectId,
    pub worktree_id: Option<WorktreeId>,
    pub relative_path: RelativePath,
    pub content: String,
    pub expected_content_hash: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub expected_modified_at_ms: Option<f64>,
    pub origin: NativeFsMutationOrigin,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct NativeFsWriteFileResult {
    pub entry: NativeFsEntry,
    pub conflict: Option<NativeFsConflict>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub file: Option<NativeFsReadFileResult>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct NativeFsConflict {
    pub reason: String,
    pub current_content_hash: Option<String>,
    pub external_mtime_ms: Option<u64>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct NativeFsCreateEntryInput {
    pub project_id: ProjectId,
    pub worktree_id: Option<WorktreeId>,
    pub parent_relative_path: Option<RelativePath>,
    pub name: String,
    pub kind: NativeFsEntryKind,
    pub origin: NativeFsMutationOrigin,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct NativeFsRenameEntryInput {
    pub project_id: ProjectId,
    pub worktree_id: Option<WorktreeId>,
    pub relative_path: RelativePath,
    pub next_name: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub next_parent_relative_path: Option<RelativePath>,
    pub origin: NativeFsMutationOrigin,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct NativeFsDeleteEntryInput {
    pub project_id: ProjectId,
    pub worktree_id: Option<WorktreeId>,
    pub relative_path: RelativePath,
    pub origin: NativeFsMutationOrigin,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct NativeFsCopyEntriesInput {
    pub project_id: ProjectId,
    pub worktree_id: Option<WorktreeId>,
    pub source_relative_paths: Vec<RelativePath>,
    pub destination_parent_relative_path: Option<RelativePath>,
    pub origin: NativeFsMutationOrigin,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct NativeFsCopyExternalEntriesInput {
    pub project_id: ProjectId,
    pub worktree_id: Option<WorktreeId>,
    pub source_paths: Vec<String>,
    pub destination_parent_relative_path: Option<RelativePath>,
    pub origin: NativeFsMutationOrigin,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct NativeFsEntryMutationResult {
    pub kind: NativeFsEntryKind,
    pub name: String,
    pub parent_relative_path: Option<RelativePath>,
    pub relative_path: RelativePath,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub entry: Option<NativeFsEntry>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct NativeFsEntryMutationListResult {
    pub entries: Vec<NativeFsEntryMutationResult>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct NativeFsRecordExternalMutationInput {
    pub project_id: ProjectId,
    pub worktree_id: Option<WorktreeId>,
    pub relative_paths: Vec<RelativePath>,
    pub origin: NativeFsMutationOrigin,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct NativeFsRevealEntryInfoInput {
    pub project_id: ProjectId,
    pub worktree_id: Option<WorktreeId>,
    pub relative_path: Option<RelativePath>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct NativeFsRevealEntryInfoResult {
    pub project_id: ProjectId,
    pub worktree_id: Option<WorktreeId>,
    pub path: String,
    pub relative_path: Option<RelativePath>,
    pub exists: bool,
    pub kind: NativeFsEntryKind,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct NativeFsWatchInput {
    pub project_id: ProjectId,
    pub worktree_id: Option<WorktreeId>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct NativeFsWatchSyncRegistryInput {
    pub projects: Vec<crate::projects::NativeProjectSummary>,
    pub worktrees: Vec<crate::projects::NativeWorktreeSummary>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct NativeProjectTreeInvalidation {
    pub project_id: ProjectId,
    pub worktree_id: Option<WorktreeId>,
    pub relative_paths: Option<Vec<RelativePath>>,
    pub occurred_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct NativeFsWatchEvent {
    pub project_id: ProjectId,
    pub worktree_id: Option<WorktreeId>,
    pub relative_path: Option<RelativePath>,
    pub kind: String,
    pub origin: NativeFsMutationOrigin,
    pub occurred_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct NativeOpenBufferState {
    pub path: String,
    pub relative_path: RelativePath,
    pub project_id: ProjectId,
    pub worktree_id: Option<WorktreeId>,
    pub content_hash: Option<String>,
    pub is_dirty: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct NativePathPolicy {
    pub path: String,
    pub visible: bool,
    pub ai_access: bool,
    pub reason: Option<String>,
}
